### PR TITLE
Feature/ca 22228/shopify push fulfillments service migration

### PR DESCRIPTION
**Pull request by clericuzzi**
In case we have a fulfillment associated with a fulfillmentOrder without supported action, we'd throw a `ShopifyUnsupportedActionException`

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -4,7 +4,7 @@
 	<modelVersion>4.0.0</modelVersion>
 	<groupId>com.channelape</groupId>
 	<artifactId>shopify-sdk</artifactId>
-	<version>2.6.0-SNAPSHOT</version>
+	<version>2.7.0-SNAPSHOT</version>
```
Comment (ryankazokas) [inline pom.xml:7]: You can leave as 2.6.0-SNAPSHOT. That's what snapshots are for. They aren't final. It only becomes final once merged into maven central without the '-SNAPSHOT'

File: src/main/java/com/shopify/ShopifySdk.java [hunk 2]
```diff
@@ -735,11 +736,18 @@ public ShopifyFulfillment createFulfillment(
...
 	 */
 	public ShopifyFulfillment createFulfillment(
 			final ShopifyFulfillmentCreationRequest shopifyFulfillmentCreationRequest,
-			List<ShopifyFulfillmentOrder> fulfillmentOrders) {
-		return this.createFulfillmentWithFulfillmentOrderApi(shopifyFulfillmentCreationRequest, fulfillmentOrders);
+			List<ShopifyFulfillmentOrder> fulfillmentOrders) throws ShopifyUnsupportedActionException {
+		try {
+			return this.createFulfillmentWithFulfillmentOrderApi(shopifyFulfillmentCreationRequest, fulfillmentOrders);
+		} catch (Exception e) {
+			throw e;
```
Comment (ryankazokas) [inline src/main/java/com/shopify/ShopifySdk.java:749]: This try catch isn't adding any context here. You can remove it, and just let the client handle the exception.

File: src/main/java/com/shopify/mappers/LegacyToFulfillmentOrderMapping.java [hunk 3]
```diff
@@ -118,6 +120,8 @@ public static ShopifyFulfillmentPayloadRoot toShopifyFulfillmentPayloadRoot(fina
 					}
 				}
 			}
+			if (lineItemsByFulfillmentOrder.size() < 1)
+				throw new ShopifyUnsupportedActionException(SupportedActions.CREATE_FULFILLMENT);
```
Comment (ryankazokas) [inline src/main/java/com/shopify/mappers/LegacyToFulfillmentOrderMapping.java:124]: I'm fine with leaving this as an exception if you want. it's a little confusing here though because it's not shopify stating that we can't do that. (We actually can if we have the right data passed in here). It's actually the input value that is preventing the action here. My goto usually if i don't want to write my own exception would be to throw an IllegalStateException that throws with the error message of something like: 'You can't create a fulfillment without any line items'. I think this would make things a bit clear to the end user.

File: src/main/java/com/shopify/model/ShopifyFulfillmentOrder.java [hunk 4]
```diff
@@ -413,4 +413,15 @@ public void setUpdatedAt(DateTime updatedAt) {
 		this.updatedAt = updatedAt;
 	}
 
+	public boolean hasSupportedAction(final SupportedActions action) {
```
Comment (ryankazokas) [inline src/main/java/com/shopify/model/ShopifyFulfillmentOrder.java:416]: You can simplify this whole function down to one line:

> return this.getSupportedActions().anyMatch(supportedAction -> supportedAction.equalsIgnoreCase(action.toString()));

It's why i love functional programming

File: src/test/java/com/shopify/ShopifySdkTest.java [hunk 5]
```diff
@@ -303,28 +304,75 @@ public void givenSomeShopifyFulfillmenmtCreationRequestWhenCreatingShopifyFulfil
...
+		fulfillmentOrderLineItem.setFulfillableQuantity(1);
+		fulfillmentOrderLineItem.setFulfillmentOrderId(fulfillmentOrderId);
+		fulfillmentOrderLineItems.add(fulfillmentOrderLineItem);
+
+		final List<String> supportedActions = new LinkedList<>();
+		supportedActions.add("move");
+
+		final ShopifyFulfillmentOrder fulfillmentOrder = new ShopifyFulfillmentOrder();
+		fulfillmentOrder.setId(fulfillmentOrderId);
+		fulfillmentOrder.setLineItems(fulfillmentOrderLineItems);
```
Comment (ryankazokas) [inline src/test/java/com/shopify/ShopifySdkTest.java:331]: Can you add a test here to show the case where fulfillmentOrder lineitems list is empty? I don't see a test here for that?

**Review by ryankazokas — APPROVED** [review]
Just some minor feedback around tests and the exception we are throwing when lineItems are empty i think should be it's own type of exception.

File: src/test/java/com/shopify/ShopifySdkTest.java (hunk 5 above)
Comment (clericuzzi) [reply src/test/java/com/shopify/ShopifySdkTest.java:331]: yep, there isn't one
the one I added acts as one because the fulfillmentOrder has no supported actions so in the end the result will be the same
but it'll be good to add one more with the conditions reversed

File: src/main/java/com/shopify/ShopifySdk.java (hunk 2 above)
Comment (clericuzzi) [reply src/main/java/com/shopify/ShopifySdk.java:749]: yea, this was me testing I was having problems catching the exception and forgot this here
tks

File: src/main/java/com/shopify/mappers/LegacyToFulfillmentOrderMapping.java (hunk 3 above)
Comment (clericuzzi) [reply src/main/java/com/shopify/mappers/LegacyToFulfillmentOrderMapping.java:124]: It's because in the case we got, the empty lineItems are caused by a lack of supported actions. But it does makes sense to change this one into a clearer message to the final user

File: src/main/java/com/shopify/model/ShopifyFulfillmentOrder.java (hunk 4 above)
Comment (clericuzzi) [reply src/main/java/com/shopify/model/ShopifyFulfillmentOrder.java:416]: I'm still somewhat confused with Java's arrow functions, I stored on my snippets here
you'll see it in my next commits

File: src/test/java/com/shopify/ShopifySdkTest.java (hunk 5 above)
Comment (ryankazokas) [reply src/test/java/com/shopify/ShopifySdkTest.java:331]: Yea, it'd be good for regression to have both cases covered for if not anything the regression we get. So future changes don't break what we expect is happening.
